numalogic/registry/dynamodb_registry.py

```diff
@@ -0,0 +1,398 @@
+                return self.cache_registry.delete(key)
+            return self.cache_registry.clear()
+        return None
+
+    @staticmethod
+    def _unpack_artifact_data(
+        item: dict[str, Any],
+    ) -> ArtifactData:
+        serialized_artifact = item.get("artifact")
+        artifact_version = item.get("version")
+        artifact_timestamp = item.get("timestamp")
+        serialized_metadata = item.get("metadata")
+
+        serialized_artifact = (
+            serialized_artifact.value
+            if hasattr(serialized_artifact, "value")
+            else serialized_artifact
+        )
+        deserialized_artifact = loads(serialized_artifact)
+        deserialized_metadata = None
+        if serialized_metadata:
+            serialized_metadata = (
+                serialized_metadata.value
+                if hasattr(serialized_metadata, "value")
+                else serialized_metadata
+            )
+            deserialized_metadata = loads(serialized_metadata).get("metadata")
+
+        return ArtifactData(
+            artifact=deserialized_artifact,
+            metadata=deserialized_metadata,
+            extras={
+                "timestamp": float(artifact_timestamp),
+                "version": artifact_version,
+            },
+        )
+
+    @staticmethod
+    def _pack_artifact_data(
+        artifact: artifact_t,
+        version: str,
+        **metadata: META_VT,
+    ) -> dict[str, Any]:
+        serialized_artifact = dumps(artifact)
+        serialized_metadata = ""
+        if metadata:
+            serialized_metadata = dumps(deserialized_object=metadata)
+
+        return {
+            "artifact": serialized_artifact,
+            "metadata": serialized_metadata,
+            "version": str(version),
+            "timestamp": str(time.time()),
+        }
+
+    def load(
+        self,
+        skeys: KEYS,
+        dkeys: KEYS,
+        latest: bool = True,
+        version: Optional[str] = None,
+    ) -> Optional[ArtifactData]:
+        """
+        Loads the desired artifact and metadata from dynamodb.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            latest: ignored for now
+            version: ignored for now
+        Returns:
+            ArtifactData instance.
+        """
+        if (latest and version) or (not latest and not version):
+            raise ValueError("Either One of 'latest' or 'version' needed in load method call")
+
+        table = self.connector.get_resource().Table(self.table_name)
+        part_key = self.construct_key(skeys)
+        sort_key = self.construct_key(dkeys)
+
+        if latest:
+            v_sort_key = self._construct_version_key(sort_key)
+        else:
+            # Version is provided
+            v_sort_key = self._construct_version_key(sort_key, version=version)
+
+        cached_artifact = self._load_from_cache(f"{part_key}::{v_sort_key}")
```

Comment:
Question: Do we want to save version provided calls to be saved in cache too?

Reply:
This already handles the version as part of the v_sort_key - the purpose of the cache is to reduce the multiple calls to the DynamoDB, if we requested a model with a version once, and then request it again, it will come from the cache and avoid the network call.

Reply:
@mboussarov when do you think we would need to call a specific version in a realtime pipeline?